src/index.c

```diff
@@ -2647,30 +2645,32 @@ static int write_disk_entry(git_filebuf *file, git_index_entry *entry, const cha
 	 *
 	 * In 2038 I will be either too dead or too rich to care about this
 	 */
-	ondisk->ctime.seconds = htonl((uint32_t)entry->ctime.seconds);
-	ondisk->mtime.seconds = htonl((uint32_t)entry->mtime.seconds);
-	ondisk->ctime.nanoseconds = htonl(entry->ctime.nanoseconds);
-	ondisk->mtime.nanoseconds = htonl(entry->mtime.nanoseconds);
-	ondisk->dev = htonl(entry->dev);
-	ondisk->ino = htonl(entry->ino);
-	ondisk->mode = htonl(entry->mode);
-	ondisk->uid = htonl(entry->uid);
-	ondisk->gid = htonl(entry->gid);
-	ondisk->file_size = htonl((uint32_t)entry->file_size);
+	ondisk.ctime.seconds = htonl((uint32_t)entry->ctime.seconds);
+	ondisk.mtime.seconds = htonl((uint32_t)entry->mtime.seconds);
+	ondisk.ctime.nanoseconds = htonl(entry->ctime.nanoseconds);
+	ondisk.mtime.nanoseconds = htonl(entry->mtime.nanoseconds);
+	ondisk.dev = htonl(entry->dev);
+	ondisk.ino = htonl(entry->ino);
+	ondisk.mode = htonl(entry->mode);
+	ondisk.uid = htonl(entry->uid);
+	ondisk.gid = htonl(entry->gid);
+	ondisk.file_size = htonl((uint32_t)entry->file_size);
 
-	git_oid_cpy(&ondisk->oid, &entry->id);
+	git_oid_cpy(&ondisk.oid, &entry->id);
 
-	ondisk->flags = htons(entry->flags);
+	ondisk.flags = htons(entry->flags);
 
 	if (entry->flags & GIT_IDXENTRY_EXTENDED) {
-		struct entry_long *ondisk_ext;
-		ondisk_ext = (struct entry_long *)ondisk;
-		ondisk_ext->flags_extended = htons(entry->flags_extended &
+		struct entry_long ondisk_ext;
+		memcpy(&ondisk_ext, &ondisk, sizeof(struct entry_short));
+		ondisk_ext.flags_extended = htons(entry->flags_extended &
 			GIT_IDXENTRY_EXTENDED_FLAGS);
-		path = ondisk_ext->path;
+		memcpy(mem, &ondisk_ext, sizeof(struct entry_long));
```

Comment:
This is in fact writing `ondisk_ext.path` into `mem`, which is not initialized yet. While I think we're fine here as `path` gets `memcpy`d to anyway, you could also use `offsetof(struct entry_long, path)` instead.

Reply:
I'm not 100% sure what you mean. Use the offset of ```path``` for what? Why shouldn't ```mem``` not be initialized? It's zero'd in line 2636, isn't it?

Reply:
`mem` is initialized, yes. But `ondisk_ext.path` is not initialized, as you only copy `sizeof(struct entry_short)` bytes into `ondisk_ext`, leaving additional struct entries (which happen to be only `path` and the extended flags) uninitialized.

Reply:
And sorry for being unclear: use `offsetof(struct entry_long, path)` as the third argument of `memcpy`

Reply:
Ah,  now I understand what you mean. Thanks for the additional details. Yes, you're right, ```path``` not initialized yet. I will use your suggested change.

Reply:
Ah, and shouldn't it be ```offsetof(struct entry_short, path)``` for both cases? We're only filling up the extended flags after the ```memcpy()``` aren't we?

Reply:
Wait, no. I looked at the wrong line. Never mind ;).